presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/PlanChecker.java

```diff
@@ -34,42 +36,39 @@
  */
 public final class PlanChecker
 {
-    private final PlanCheckerProviderManager providerMgr;
-    private Multimap<Stage, Checker> checkers;
-    private boolean providedInitialized;
+    @GuardedBy("this")
+    private volatile Multimap<Stage, Checker> checkers;
```

Comment:
Synchronization here can be a performance problem when there are a lot of concurrent queries running. Since we're only updating once, make this a concurrentHashMap with a CopyOnWriteArrayList for the values, so you don't need to synchronize access. Also, make this final, and then you can remove the volatile keyword.

as an alternative, you could skip adding any plugin plan checkers to this map completely, and instead call to the PlanCheckerProvidermanager when you do the checks to get any extra plan checkers that they provide. (the suggestion from https://github.com/prestodb/presto/pull/23596#discussion_r1783477665)

Reply:
I do prefer the second option, since it would just be iterating over the list of provided checkers so it wouldn't be adding overhead and it would certainly be cleaner. If I understand correctly this would make a check look something like this:

```java
    public synchronized void validateFinalPlan(PlanNode planNode, Session session, Metadata metadata, WarningCollector warningCollector)
    {
        checkers.get(Stage.FINAL).forEach(checker -> checker.validate(planNode, session, metadata, warningCollector));
        for (PlanCheckerProvider provider : planCheckerProviderManager.getPlanCheckerProviders()) {
            for (com.facebook.presto.spi.plan.PlanChecker checker : provider.getPlanCheckersFinal()) {
                checker.validate(planNode, warningCollector);
            }
        }
    }
``` 

That would require the `PlanChecker` keep the instance to the `PlanCheckerProviderManager`. I think @tdcmeehan preferred to not have that, so I'd like him to weigh in.

Reply:
I don't think it occurred to me when I made my suggestion that even the read-only methods would need to be synchronized.

Re-reading this code, I'm wondering if we even need `PlanCheckerManager`, and if instead we should just do option 1 and have the existing `PlanChecker` class in `presto-main` subsume all of its responsibilities.  It seems the `PlanCheckerManager` is just an intermediary between the two, and it seems better consolidate it.

Reply:
I just now remembered that there's a single node and distributed instance of `PlanChecker`, and that depends on a toggle that's known at runtime.  Agreed that option 2 looks like the best approach!